fix #514, call_data_callbacks after shared the data

jmorice91 opened the pull request:
!!!INSERT YOUR DESCRIPTION HERE!!!

# List of things to check before making a PR

Before merging your code, please check the following:

* [x] you have added a line describing your changes to the Changelog;
* [x] you have added unit tests for any new or improved feature;
* [x] In case you updated dependencies, you have checked pdi/docs/CheckList.md
* you have checked your code format:
  - [x] you have checked that you respect all conventions specified in CONTRIBUTING.md;
  - [x] you have checked that the indentation and formatting conforms to the `.clang-format`;
  - [x] you have documented with doxygen any new or changed function / class;
* you have correctly updated the copyright headers:
  - [x] your institution is in the copyright header of every file you (substantially) modified;
  - [x] you have checked that the end-year of the copyright there is the current one;
* you have updated the AUTHORS file:
  - [x] you have added yourself to the AUTHORS file;
  - [x] if this is a new contribution, you have added it to the AUTHORS file;
* you have added everything to the user documentation:
  - [x] any new CMake configuration option;
  - [x] any change in the yaml config;
  - [x] any change to the public or plugin API;
  - [x] any other new or changed user-facing feature;
  - [x] any change to the dependencies;
* you have correctly linked your MR to one or more issues:
  - [x] your MR solves an identified issue;
  - [x] your commit contain the `Fix #issue` keyword to autoclose the issue when merged.

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +143,31 @@ void warn_status(PDI_status_t status, const char* message, void*)
+struct Var_to_reclaim {
+	list<string> names;
+
+	Var_to_reclaim() { names.clear(); }
```
```suggestion
	Var_to_reclaim() = default;
```

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +143,31 @@ void warn_status(PDI_status_t status, const char* message, void*)
 	}
 }
 
+struct Var_to_reclaim {
```
Missing documentation

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +143,31 @@ void warn_status(PDI_status_t status, const char* message, void*)
+
+	Var_to_reclaim() { names.clear(); }
+
+	~Var_to_reclaim()
```
Can the destructor throw ? If so, how is the exception handled ?

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -324,37 +349,36 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, void* da
+			void* v_data = va_arg(ap, void*);
+			PDI_inout_t v_access = static_cast<PDI_inout_t>(va_arg(ap, int));
+
+			Global_context::context().logger().trace("\n Multi expose: Sharing `{}' ({}/{}) \n", ++i, list_reclaim.names.size());
```
```suggestion
			Global_context::context().logger().trace("Multi expose: Sharing `{}' ({}/{})", v_name, ++i, list_reclaim.names.size());
```
There is an argument missing here.

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -324,37 +349,36 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, void* da
-		if ((status = PDI_share(v_name, v_data, v_access))) {
-			break;
+
+	Var_to_reclaim list_reclaim;
```
Add comment to explain what this does.

benoitmartin88 commented on pdi/src/data_descriptor_impl.h:
```diff
@@ -89,9 +89,11 @@ class PDI_EXPORT Data_descriptor_impl: public Data_descriptor
-	void* share(Ref ref, bool read, bool write) override;
+	void* share(Ref ref, bool read, bool write, bool delayDataCallback = false) override;
+
+	void data_callbacks() override;
```
Is this new method really needed ?
It is only used once. If the goal is to factorize calls to `call_data_callbacks`, than it should be used everywhere where `call_data_callbacks` is used.

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -324,37 +349,36 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, void* da
-			break;
+
+	Var_to_reclaim list_reclaim;
+	{
```
```suggestion
```
I don't think this is needed

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -324,37 +349,36 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, void* da
+			Global_context::context().logger().trace("Multi expose: data events `{}' ({}/{})", it->c_str(), ++i, list_reclaim.size());
+			Global_context::context()[it->c_str()].data_callbacks();
+		}
 	}
```
```suggestion
```

benoitmartin88 replied:
```suggestion
```

benoitmartin88 reviewed (changes requested):
Great feature to add. Thanks !
Not much is needed to merge this PR. 
I think this feature would require a unit test. Could you add this?

benoitmartin88 reviewed (changes requested):
Almost ready to merge.
I do have a few questions that need answering concerning error handling, potentially superfluous functions and writing tests in C++.

benoitmartin88 commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -151,13 +151,30 @@ bool Data_descriptor_impl::empty()
+	throw Error(e.status(), "Unable to execute data_callbacks on data `{}', {}", name(), e.what());
+}
+
+
```
```suggestion
```

benoitmartin88 commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -214,6 +228,7 @@ try {
 	throw Error(e.status(), "Unable to share `{}', {}", name(), e.what());
 }
 
+
```
```suggestion
```

benoitmartin88 reviewed (changes requested):
I think this is almost ready to merge. 
Could you make sure the indent and spacing of the modified files is correct ?
Thanks !

Yushan-Wang reviewed (approved):
Ready to merge

jbigot commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +143,29 @@ void warn_status(PDI_status_t status, const char* message, void*)
 
+/** A structure to reclaim the datas properly in case of error
+ */
+struct Var_to_reclaim: public std::list<string> {
```
I really like the idea. This seems to go the way of what was proposed above. I would like it to be be public and used on the provider side rather than on the user side though.

Also, as a side note, inheriting from a `std::list` is not something I would do. I'd rather have it as a member. Cf. https://en.wikipedia.org/wiki/Composition_over_inheritance

jbigot reviewed (changes requested):
This is a great submission! My main comment is about moving `Var_to_reclaim` of the provider side instead of expecting each user to implement it. Overall this is really nice though!

benoitmartin88 commented on pdi/include/pdi/context.h:
```diff
@@ -33,6 +33,7 @@
 
 #include <pdi/pdi_fwd.h>
 #include <pdi/callbacks.h>
+//#include <pdi/delayed_data_callbacks.h>
```
Why is this commented ?

benoitmartin88 commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,150 @@
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks, {}", e.what());
+			} else {
+				// No exception is throwing. Throw an error.
+				throw Error(e.status(), "Error in the destructor of Delayed_data_callbacks, {}", e.what());
```
this will fail. You are declaring your destructor as noexcept, and here you are throwing.

Yushan-Wang commented on pdi/include/pdi/context.h:
```diff
@@ -33,6 +33,7 @@
 
 #include <pdi/pdi_fwd.h>
 #include <pdi/callbacks.h>
+//#include <pdi/delayed_data_callbacks.h>
```
remove commented-out lines

jmorice91 replied:
This will be done in the next commit

benoitmartin88 commented on pdi/src/data_descriptor_impl.h:
```diff
@@ -43,10 +44,13 @@ namespace PDI {
+
 	struct PDI_NO_EXPORT Ref_holder;
 
+public:
```
why public ?

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +144,34 @@ void warn_status(PDI_status_t status, const char* message, void*)
+/** A structure to reclaim the datas properly in case of error
+ */
+struct Var_to_reclaim {
+	std::list<string> m_varnames;
```
I think that a vector here makes more sense.

jbigot replied:
we should go for a vector unless there is a good reason to favor otherwise

benoitmartin88 commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +144,34 @@ void warn_status(PDI_status_t status, const char* message, void*)
+	std::list<string> m_varnames;
+	Var_to_reclaim() = default;
+
+	~Var_to_reclaim()
```
```suggestion
	~Var_to_reclaim() noexcept
```

benoitmartin88 reviewed (changes requested):
Thank you for this contribution. There are a few changes that are required.

Yushan-Wang commented on pdi/tests/PDI_callbacks.cxx:
```diff
@@ -649,3 +651,151 @@ TEST_F(CallbacksTest, add_data_remove_callback_release_remove)
+			Ref_w ref_write{ref};
+			int* x = static_cast<int*>(ref_write.get());
+			*x += 42;
+			std::cout << "name=" << name.c_str() << " == data_x" << std::endl;
```
remove stdcout and iostream

jmorice91 committed:
first commit multi_expose delay

add Delayed_data_callbacks class

remove cout

Adding some test how work delayed_data_callbacks class

fix comment

fix indent and word

jmorice91 committed:
first commit multi_expose delay

add Delayed_data_callbacks class

Adding some test how work delayed_data_callbacks class

Yushan-Wang commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -30,9 +30,14 @@
 #include <pdi/datatype_template.h>
 #include <pdi/ref_any.h>
 
+// #include "global_context.h"
```
remove?

jmorice91 replied:
I will be remove in the next commit

jmorice91 commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,150 @@
+			} else {
+				// No exception is throwing. Throw an error.
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks.");
+				throw;
```
same comment as above => close this conversation

jmorice91 commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,150 @@
+			} else {
+				// No exception is throwing. Throw an error.
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks.");
+				throw;
```
In this choice of implementation, the destructor may throw. 
It is for this reason, we  add `noexcept(false)` to the definition of the dtor.

jbigot commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -30,9 +30,14 @@
 #include <pdi/ref_any.h>
 
+// #include "global_context.h"
+
```
```suggestion
```

jbigot commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -85,6 +90,14 @@ class PDI_EXPORT Data_descriptor
+	 * \param[in,out] data the shared data
+	 * \param read whether read access is granted to other references
+	 * \param write whether write access is granted to other references
+	 * \param delayed_callbacks a class that allow to delay trigger_delayed_data_callbacks for a list of a data
```
```suggestion
	 * \param delayed_callbacks a list of callbacks where the callback for this data will be added, instead of being triggered. So that one can delay the trigger
```

jbigot commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -93,6 +106,19 @@ class PDI_EXPORT Data_descriptor
+	 * \param[in,out] ref a reference to the shared data
+	 * \param read whether the stored reference should have read access
+	 * \param write whether the stored reference should have write access
+	 * \param delayed_callbacks a class that allow to delay trigger_delayed_data_callbacks for a list of a data
```
```suggestion
	 * \param delayed_callbacks a list of callbacks where the callback for this data will be added, instead of being triggered. So that one can delay the trigger
```

jbigot commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -93,6 +106,19 @@ class PDI_EXPORT Data_descriptor
+
+	/** function to call "data_callbacks" for the shared data.
+	 */
+	virtual void trigger_delayed_data_callbacks() = 0;
```
Why is it still required?

jmorice91 replied:
We need that because `m_context[element_name.c_str()]` is  a `data_descriptor` in **Delayed_data_callbacks**

Do you want to remove **trigger_delayed_data_callbacks()** in the class `data_descriptor_impl` and put this function inside  `class PDI_EXPORT Delayed_data_callbacks`?

I want to keep this function inside `data_descriptor_impl`  because I don't want to change `m_refs` outside this class
```c
/// References to the values of this descriptor
std::stack<std::unique_ptr<Ref_holder>> m_refs;
```

jmorice91 replied:
This function will be remove after discussion.

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+
+class PDI_EXPORT Delayed_data_callbacks
+{
+	/// friend class Global_context;
```
What does this comment mean?

jmorice91 replied:
this will be deleted

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+	/// friend class Global_context;
+	std::vector<std::string> m_datanames;
+
+	/// The context this descriptor is part of
```
This is not a descriptor, is it?

jmorice91 replied:
I fix this error in the next commit

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+public:
+	/// constructor used in:
+	///     - the case of PDI_multi_expose
+	///     - when a user want to define is own "Delayed_data_callbacks" object
```
```suggestion
	///     - when a user want to define its own "Delayed_data_callbacks" object
```

jmorice91 replied:
In previous implementation, I have two contructor for this objet. Now, I have only. 
Perhaps, it is better to remove this comment

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+	/// constructor used in:
+	///     - the case of PDI_multi_expose
+	///     - when a user want to define is own "Delayed_data_callbacks" object
+	/// Remark: This constructor doesn't work if PDI_init is not called.
```
```suggestion
	/// \remark: This constructor doesn't work if PDI_init is not called.
```

jmorice91 replied:
In the test of 'data descriptor' object, I have encounter this issue. As we fixed this problem, this comment must be irrelevant.

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks, {}", e.what());
+			} else {
+				// No exception is throwing. Throw an error.
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks.");
```
since you rethrow. I wouldn't print anything in that case:

```suggestion
```

jmorice91 replied:
done in the next commit

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks.");
+			} else {
+				// No exception is throwing. Throw an error.
+				m_context.logger().error("Error in the destructor of Delayed_data_callbacks.");
```
same as above

```suggestion
```

jmorice91 replied:
will be done in the next commit

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+	//  (example: error in the config.yml for a plugin, error due to external library incompatibility)
+	~Delayed_data_callbacks() noexcept(false)
+	{
+		try {
```
why not move the try one step up?

jmorice91 commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,150 @@
+		try {
+			this->trigger();
+		} catch (Error& e) {
+			if (std::uncaught_exceptions()) {
```
The paper in reference talk about `bool std::uncaught_exception` (Attention: there is no `s`at the end).
It recommend to introduce `int std::uncaught_exceptions()` which is used here (introduced in c++17).

If we have a `throw` in other part of the code,  `std::uncaught_exceptions()` allows to know if the stack unwinding is currently in process. This **if** condition is used to avoid throwing in the destructor in this case.

jmorice91 commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,150 @@
+				throw;
+			}
+		} catch (...) {
+			if (std::uncaught_exceptions()) {
```
same answer as above => close

jmorice91 commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,150 @@
+				throw Error(e.status(), "Error in the destructor of Delayed_data_callbacks, {}", e.what());
+			}
+		} catch (const std::exception& e) {
+			if (std::uncaught_exceptions()) {
```
same answer as above => close

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,128 @@
+	} catch (...) {
+		this->cancel(); // clear the m_datanames
+		throw;
+	}
```
When an exception is thrown, we'd like the events not yet triggered to remain in the list of events to trigger.

```suggestion
	void trigger()
	{
		int i = 0;
		size_t number_of_elements = m_datanames.size();
		while ( ! m_datanames.empty() ) {
		    auto&& element_name = m_datanames.back();
			m_context.logger().trace("Trigger data callback `{}' ({}/{})", element_name, ++i, number_of_elements);
			m_context.callbacks().call_data_callbacks(element_name, m_context[element_name].ref());
			m_datanames.pop_back()
		}
	}
```

jbigot commented on pdi/src/pdi.cxx:
```diff
@@ -143,6 +143,34 @@ void warn_status(PDI_status_t status, const char* message, void*)
 
+/** A structure to reclaim the datas properly in case of error
+ */
+struct Var_to_reclaim {
```
Why is this class still required? Isn't `Delayed_data_callbacks` enough?

jmorice91 replied:
These two classes are defined with the same list of dataname in some sense. 
(It depends only if we have an error after `delayed_callbacks.add_dataname(m_name);` 
 in `void* Data_descriptor_impl::share`) -> good catch.

Remark:
In **PDI_share**, the destruction of the object  `Delayed_data_callbacks` execute event `on_data` in this implementation.
The class `Var_to_reclaim` is used to call **PDI_reclaim** the data in **PDI_multi_expose**.

So if we want to have only 1 class:
- need to rename the class `Delayed_data_callbacks` -> name ???
- add a member to this class to know if we are in the **multi_expose** case to have two version of destructor:
1.  dtor1 : call trigger()
2.  dtor2:  call trigger() + reclaim the variables.
- Perhaps add a member to this class to know if we call trigger() before.
- don't remove element in the list of the dataname in trigger()

Is it what you want to implement? we can discuss if you want.

jbigot committed:
Improved error messages for configuration errors

Also simplified error code

Fix #657

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -151,13 +152,47 @@ bool Data_descriptor_impl::empty()
+	try {
+		m_context.callbacks().call_data_callbacks(m_name, ref());
+	} catch (const exception&) {
+		m_refs.pop();
```
Why pop in that case? If, as I suspect, it's not needed, I would just remove all the exception handling code here.

jmorice91 replied:
Okay

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -151,13 +152,47 @@ bool Data_descriptor_impl::empty()
+	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");
+	if (m_refs.empty() || (m_refs.size() == 1 && metadata())) {
+		throw State_error{"Cannot call trigger_delayed_data_callbacks on a non shared value: : `{}'", m_name};
+	}
```
Isn't this already checked in the call to `ref()`. If so, I wouldn't re-check here.

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -170,6 +205,17 @@ try {
+	return share(data_ref, read, write, Delayed_data_callbacks(m_context));
+} catch (Error& e) {
+	throw Error(e.status(), "Unable to share `{}', {}", name(), e.what());
+	;
```
```suggestion
```

jmorice91 replied:
okay

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -201,12 +247,7 @@ try {
-	try {
-		m_context.callbacks().call_data_callbacks(m_name, ref());
-	} catch (const exception&) {
-		m_refs.pop();
```
the reason we poped here is so that the call would be a no-op if an exception was thrown in the callback. This doesn't work anymore with delayed callback. I would instead turn callback exceptions into warning instead

jmorice91 replied:
I understand now

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -151,13 +152,47 @@ bool Data_descriptor_impl::empty()
+try {
+	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");
+
+	share(data, read, write, Delayed_data_callbacks(m_context));
+
+	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");
+} catch (Error& e) {
+	throw;
+}
```
no point in a try-catch where you just re-throw; that's the default behaviour

```suggestion
{
	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");

	share(data, read, write, Delayed_data_callbacks(m_context));

	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");
}
```

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -170,6 +205,17 @@ try {
+	return share(data_ref, read, write, Delayed_data_callbacks(m_context));
+} catch (Error& e) {
+	throw Error(e.status(), "Unable to share `{}', {}", name(), e.what());
+	;
+} catch (...) {
+	throw std::runtime_error("void* Data_descriptor_impl::share(Ref data_ref, bool read, bool write)");
+}
```
I'm not completely sure about this. Why the different treatment from above. I'd make it a PDI::Error if you want to handle it.

Also, if it's a `std::exception`, I would re-use the message.

jbigot commented on pdi/include/pdi/delayed_data_callbacks.h:
Please move the implementation to a `.cxx` file in `src/`

jmorice91 replied:
I will do that in the next commit

jbigot commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -151,13 +152,47 @@ bool Data_descriptor_impl::empty()
 	return m_refs.empty();
 }
 
+void Data_descriptor_impl::trigger_delayed_data_callbacks()
```
In the end, with the simplification in error management, I believe you end up with just:
```C++
m_context.callbacks().call_data_callbacks(m_name, ref());
```
So I wouldn't dedicate a function for this and make this call directly in `Delayed_data_callbacks::trigger()`

jmorice91 replied:
okay

Yushan-Wang commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -152,12 +153,21 @@ bool Data_descriptor_impl::empty()
+
+	share(data, read, write, Delayed_data_callbacks(m_context));
+
+	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");
```
Why double assert?

jmorice91 replied:
I remove this.

Yushan-Wang commented on pdi/src/data_descriptor_impl.cxx:
```diff
@@ -152,12 +153,21 @@ bool Data_descriptor_impl::empty()
+
+void Data_descriptor_impl::share(void* data, bool read, bool write, Delayed_data_callbacks&& delayed_callbacks)
 try {
 	assert((!metadata() || !m_refs.empty()) && "metadata descriptors should always keep a placeholder");
```
Do we need a third call to `assert`?

Yushan-Wang commented on pdi/src/pdi.cxx:
```diff
@@ -324,37 +380,31 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, const vo
+
+	Var_to_reclaim list_names; // list of variable that will be reclaimed at the end of this function
+	Delayed_data_callbacks delayed_callbacks(Global_context::context());
+	int i = -1;
```
Why isn't `int i=0`?

jmorice91 replied:
In the orginal code, there is no logger message for the first data exposed. As I have added 1, I need to begin my index to -1 to be consistent with the old implementation.

JAuriac commented on pdi/include/pdi/pdi_fwd.h:
```diff
@@ -79,6 +79,11 @@ using Datatype_template_sptr = std::shared_ptr<const Datatype_template>;
 using Datatype_sptr = std::shared_ptr<const Datatype>;
 
+/**
+ * A class to delayed the moment to call call_data_callback
```
```suggestion
 * A class to delay the moment to call call_data_callback
```

JAuriac commented on pdi/src/delayed_data_callbacks.cxx:
```diff
@@ -0,0 +1,122 @@
+	, m_context{ctx}
+{}
+
+// In the destructor, we need to throw an error message in case of the callback on the data doesn't work (trigger function)
```
```suggestion
// In the destructor, we need to throw an error message in case the callback on the data doesn't work (trigger function)
```

JAuriac commented on pdi/src/pdi.cxx:
```diff
@@ -324,37 +392,31 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, const vo
+	Global_context::context().logger().trace("Multi expose: Calling event `{}'", event_name);
+	Global_context::context().event(event_name);
+
+	// remark: The reclaim of the datas are done in the destructor of the Var_to_reclaim (see struct Var_to_reclaim)
```
```suggestion
	// remark: The reclaim of the data is done in the destructor of the Var_to_reclaim (see struct Var_to_reclaim)
```

JAuriac commented on pdi/tests/PDI_callbacks.cxx:
```diff
@@ -649,3 +651,149 @@ TEST_F(CallbacksTest, add_data_remove_callback_release_remove)
+			int* x = static_cast<int*>(ref_write.get());
+			*x += 42;
+			std::cout << "name=" << name.c_str() << " == data_x" << std::endl;
+			ASSERT_STREQ(name.c_str(), "data_x");
```
For this file, all use of 'ASSERT' inside of a lambda function are not advisable, as they expand to `return;` on a failure, which only exit the lambda function instead of giving a failure. Prefer the use of 'EXPECT'.

JAuriac replied:
Nevermind, this is used inside of lambda functions to ensure that the equality is checked for the expected value ("data_x" in this case).

JAuriac commented on pdi/tests/PDI_callbacks.cxx:
```diff
@@ -649,3 +651,149 @@ TEST_F(CallbacksTest, add_data_remove_callback_release_remove)
+	try {
+		this->test_context->desc("data_x").share(&x, true, true, std::move(delayed_callbacks));
+		throw std::runtime_error("forced throw error for testing");
+		delayed_callbacks.trigger();
```
If this line is intended to represent the intended placement of `trigger`, it may need an inline comment to specify this (as it is never reached here).

jmorice91 replied:
the tests have been changed

Yushan-Wang reviewed (commented):
Looks good to me! 
However, I have difficulties understanding the test. 
Could you please add a test with hdf5, where the name of the file is a variable?
Before, we had to expose first the `filename_size`, then the `filename`. With your change, the order of the expose is then irrelevant. 
Or exposing a vector who's size is also a variable?

jbigot commented:
For `Multiple_error`, you could base your work on top of https://github.com/pdidev/pdi/tree/improved_errmsg
I'm waiting for the formatting & renaming of errors to be merged before I make a PR

jmorice91 committed:
tests good

jmorice91 committed:
update to the testing framework

JAuriac commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -85,6 +85,14 @@ class PDI_EXPORT Data_descriptor
 	 */
 	virtual void share(void* data, bool read, bool write) = 0;
 
+	/** Shares some data with PDI
```
```suggestion
	/** Shares some data with PDI using delay
```

JAuriac commented on pdi/include/pdi/data_descriptor.h:
```diff
@@ -93,6 +101,15 @@ class PDI_EXPORT Data_descriptor
 	 */
 	virtual void* share(Ref ref, bool read, bool write) = 0;
 
+	/** Shares some data with PDI
```
```suggestion
	/** Shares some data with PDI using delay
```

JAuriac commented on pdi/include/pdi/delayed_data_callbacks.h:
```diff
@@ -0,0 +1,66 @@
+	std::vector<std::string> m_datanames;
+
+	/// The context where the list of data is a part of
+	Global_context& m_context;
```
Maybe shouldn't use a Global_context directly, and should remove the direct include to global_context.h, to go through Context instead with virtuals

jmorice91 replied:
Good catch.  We must not use Global_context directly. This variables is not defined in the "pdi/include"
We have three options:
- 1) Use instead the class Context as you propose (implemented and tested but not commited yet).
- 2) Do a virtual class Delayed_data_callbacks and put the implementation in the src as Data_descriptor
and use `Global_context` in implementation
- 3) Do a virtual class Delayed_data_callbacks and put the implementation in the src as Data_descriptor
and use `Context` in the implementation

jmorice91 replied:
@jbigot @pdidev/pdi-developers I put this PR as draft until a choice to remove the Global_context is chosen

JAuriac commented on pdi/src/delayed_data_callbacks.cxx:
```diff
@@ -0,0 +1,103 @@
+			// An exception is throwing before. Print simple message to avoid std::terminate.
+			m_context.logger().error("Error in the destructor of Delayed_data_callbacks, {}", e.what());
+		} else {
+			m_context.logger().info("{}","#### exception:: throw the same error");
```
What does "same error" means here ?

JAuriac commented on pdi/src/delayed_data_callbacks.cxx:
```diff
@@ -0,0 +1,103 @@
+void Delayed_data_callbacks::add_dataname(const std::string& name)
+{
+	// Comment: In case of a multi_expose, if the data is defined twice then the callback "on_data" are called twice also in PDI v1.10
+	//         Therefore, we need to add the name to have the same behaviour.
```
What is the "same behaviour" here ? We need this emplace_back to reproduce the behaviour where we overwrite the data definition that is already available ?

jmorice91 replied:
If you call this function,
```c
PDI_multi_expose("my_test", "pdi_var1", &var1, PDI_OUT, "pdi_var1", &var1, PDI_OUT, NULL);
```
The callback on data "pdi_var1" is called two times in pdi v1.10.
For example, if the callback corresponds to  the user_code function  `my_func`, we call two times the function `my_func`.

JAuriac commented on pdi/src/pdi.cxx:
```diff
@@ -323,37 +394,31 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, const vo
+	Delayed_data_callbacks delayed_callbacks(Global_context::context());
+	int i = -1;
+	Global_context::context().logger().trace("Multi expose: Sharing `{}' ({}/{})", name, ++i, list_names.size());
+	Global_context::context()[name].share(const_cast<void*>(data), access & PDI_OUT, access & PDI_IN, std::move(delayed_callbacks));
```
```suggestion
	Global_context::context()[name].share(const_cast<void*>(data), access& PDI_OUT, access& PDI_IN, std::move(delayed_callbacks));
```

JAuriac commented on pdi/src/pdi.cxx:
```diff
@@ -323,37 +394,31 @@ PDI_status_t PDI_multi_expose(const char* event_name, const char* name, const vo
-		}
-		transaction_data.emplace_back(v_name);
+		Global_context::context().logger().trace("Multi expose: Sharing `{}' ({}/{})", v_name, ++i, list_names.size());
+		Global_context::context()[v_name].share(v_data, v_access & PDI_OUT, v_access & PDI_IN, std::move(delayed_callbacks));
```
```suggestion
		Global_context::context()[v_name].share(v_data, v_access& PDI_OUT, v_access& PDI_IN, std::move(delayed_callbacks));
```

jmorice91 replied:
```c
v_access & PDI_OUT
```
here, & correspond to the binary operator.

JAuriac commented on pdi/tests/PDI_delayed_data_callbacks.cxx:
```diff
@@ -0,0 +1,306 @@
+/*
+ * Name:                DataDescDelayed.multiple_data_callbacks
+ *
+ * Description:         TODO: Jacques
```
A 'TODO' remains

JAuriac commented on plugins/decl_hdf5/tests/decl_hdf5_tests.cxx:
```diff
@@ -841,8 +838,10 @@ logging: trace
+			StrEq("Invalid entry in specification tree: (13:28 -> 13:49) Unable to share `array_data', "
+				"`1' error(s) while triggering data callbacks, "
+				"while sharing `array_data', "
+				"Dataset selection is invalid for implicit dataset `group123/array_data'")
```
```suggestion
			StrEq("Invalid entry in specification tree: (13:28 -> 13:49) Unable to share `array_data', "
				  "`1' error(s) while triggering data callbacks, "
				  "while sharing `array_data', "
				  "Dataset selection is invalid for implicit dataset `group123/array_data'")
```

jmorice91 committed:
fix remarks

JAuriac commented on pdi/src/delayed_data_callbacks.cxx:
```diff
@@ -0,0 +1,101 @@
+
+	m_datanames.clear();
+
+	rethrow_with_context(trigger_data_errors, "`{}' error(s) while triggering data callbacks, ", number_of_elements);
```
This may actually represent `trigger_data_errors.size()` (is it intentional ?), as if we have 3 triggers and only 1 throws, we get a report of 3 errors

jmorice91 replied:
Here, `number_of_elements` correspond to `trigger_data_errors.size()`.

JAuriac commented on pdi/src/delayed_data_callbacks.cxx:
```diff
@@ -0,0 +1,101 @@
+			// An exception is throwing before. Print simple message to avoid std::terminate.
+			m_context.logger().error("Error in the destructor of Delayed_data_callbacks, {}", e.what());
+		} else {
+			m_context.logger().info("{}", "Error in the destructor of Delayed_data_callbacks");
```
Could use `e.what()` in the logger info branch ?

jmorice91 replied:
If I do that, in the final message, I have two times `e.what()`

JAuriac commented on tests/combination_tests/test_06_multi_expose.cxx:
```diff
@@ -0,0 +1,121 @@
+	const int var1 = 3;
+	const int var2 = 11;
+
+	PDI_multi_expose("my_test", "pdi_var1", &var1, PDI_OUT, "pdi_var1", &var2, PDI_OUT, NULL);
```
Is writing over const variables intended here (technically ok because not constexpr) ?

jmorice91 replied:
Yes, indeed I need to remove the const here.

jmorice91 committed:
fix indent

jmorice91 committed:
comment test that failed
